Automattic/portkey-android
# Color picker v2: 3 - Text color background

Part of the color picker v2 series. Builds on top of #552.

Adds a second color picker row to the bottom sheet, and wires it up to change the text color background.

![text-background-color-sheet](https://user-images.githubusercontent.com/9613966/95168082-b6c4ec80-07eb-11eb-8aba-5e07ef37e830.png)

The initial implementation used out-of-the-box `BackgroundColorSpan`, which is pretty ugly with no padding and straight edges. Per the designs we want more space around the text and rounded corners. The first few commits though use a vanilla `BackgroundColorSpan` just to showcase the base functionality. After that, I added a custom rounded background color span (mostly based on findings from SO as noted in the file).

The text color button coloring behavior will be added in the next PR.

### A few notes:
* Per the designs, fonts that have custom shadow settings have the shadow removed when a background color is set (this ended up being needed anyway, since a text shadow with the same radius as the font padding is applied with the background color).
* There are sometimes tiny, transparent artifacts visible if you zoom way in on a multi-line text.
* Under some circumstances, the background span is cut off at the end of the text. See below:

#### Text padding cut off issue

When adding text, sometimes the end highlight in the added view gets cut off a little, so the angles are sharp instead of rounded.

Editing this text and returning, or any action that pauses the fragment (like opening the media picker and returning), seems to redraw the view and correct the problem. I've tried manually invalidating the view but that doesn't seem to do it. 

![text-cutoff](https://user-images.githubusercontent.com/9613966/95169606-1e7c3700-07ee-11eb-8176-9cd012566fdc.png)


As far as I can tell, this only happens if the view contains an explicit newline (a long text that gets line broken doesn't seem to count).

I think it's better to move on to other things as I'm out of ideas, but @mzorz wondering if you have any.

### To test:
* Try applying and clearing the background color (the first option in the background color recycler view)
* Check that fonts with shadows (Casual, Playful, and Bold) always have a shadow if there's no background color, and never have a shadow when there is
* Check that cycling alignment and text style works as expected
